docs: document frozen structs (#204)
Quantinuum/guppy-docs: sphinx/language_guide/data_types/structs.md

PauliString.check()
```

## Structs are mutable and affine by default

> As mentioned, Guppy structs mutable by default. Therefore we are able to modify the fields of the struct once its defined.

Reviewer: ```suggestion
Guppy structs are mutable by default: we can modify the fields of a struct after it has been defined.
```

```{code-cell} ipython3

@guppy
def mutate_pauli() -> None:
    # Create an instance of PauliString to represent XZX (XIX * IZI ~ XZX).
    my_pauli = PauliString(array(True, False, True), array(False, True, False)) 

    # After this mutation, my_pauli now represents XIX
    my_pauli.zs = array(False, False, False)

mutate_pauli.check(); 
```

<!---
Structs are also affine by default meaning they cannot be implicitly copied with an assignment.

```{code-cell} ipython3
---
tags: [raises-exception]
---

@guppy
def implicit_copy() -> None:
    my_pauli = PauliString(array(True, False, True), array(False, True, False)) 

    my_new_pauli = my_pauli 

    # Trying to access my_pauli after a move gives an error.
    my_pauli.zs = array(False, False, False)

> implicit_copy.check(); 

Author: This should give an error but currently doesn't see https://github.com/Quantinuum/guppylang/issues/2117

Author: I've commented out this part so it doesn't appear in the built docs pages. 

I could add instead another (non-buggy) snippet which demonstrates the expected behaviour. Perhaps similar to the one I have here in the migration guide?

https://docs.quantinuum.com/guppy-v1.x/v1_migration.html#guppy-structs-are-now-mutable-by-default

Reviewer: Fine to leave for now, we can expand once the bug is fixed.

```
-->

> If we want to make a struct immutable, then we can specify the `frozen=True` keyword argument just like [Python dataclasses](https://docs.python.org/3/library/dataclasses.html#frozen-instances).

Reviewer: ```suggestion
If we want to make a struct immutable, then we can specify the `frozen=True` keyword argument, similarly to [Python dataclasses](https://docs.python.org/3/library/dataclasses.html#frozen-instances).
```
